gmso/atomtyping/parameterize.py

```diff
@@ -0,0 +1,275 @@
+"""Functions used to atomtype a gmso.Topology."""
+from operator import attrgetter
+
+import networkx as nx
+from foyer.atomtyper import AtomTypingRulesProvider, find_atomtypes
+from foyer.smarts import SMARTS
+from foyer.topology_graph import TopologyGraph
+
+from gmso import Atom
+from gmso.atomtyping.isomorph import (  # check refs
+    partition_isomorphic_topology_graphs,
+)
+
+__all__ = ["apply"]
+
+
+def apply(
+    top,
+    forcefields,
+    identify_connected_components=True,
+    use_residue_info=False,
+    assert_bond_params=True,
+    assert_angle_params=True,
+    assert_dihedral_params=True,
+    assert_improper_params=False,
+):
+    """Set Topology parameter types from GMSO Forcefields.
+
+    Parameters
+    ----------
+    top: gmso.core.topology.Topology, required
+        The GMSO topology on which to apply forcefields
+    forcefields: dict, required
+        The keys are labels that match the subtopology label or site residue_name, and the
+        values are gmso Forcefield objects that gets applied to the specified molecule
+    identify_connected_components: bool, optional, default=True
+        A flag to determine whether or not to search the topology for repeated disconnected
+        structures, otherwise known as molecules and type each molecule only once.
+    use_residue_info: bool, optional, default=False
+        A flag to determine whether or not to look at site.residue_name to look parameterize
+        each molecule only once. Will only be used if identify_connected_components=False
+    assert_bond_params : bool, optional, default=True
+        If True, Foyer will exit if parameters are not found for all system
+        bonds.
+    assert_angle_params : bool, optional, default=True
+        If True, Foyer will exit if parameters are not found for all system
+        angles.
+    assert_dihedral_params : bool, optional, default=True
+        If True, Foyer will exit if parameters are not found for all system
+        proper dihedrals.
+    assert_improper_params : bool, optional, default=False
+        If True, Foyer will exit if parameters are not found for all system
+        improper dihedrals.
+    """
+    top.identify_connections()
+    top_graph = get_modified_topology_graph(top)
+    if identify_connected_components:  # True, (True,False)
+        # populate molecule information
+        isomorphic_top_graphs = partition_isomorphic_topology_graphs(top_graph)
+        apply_atomtypes_from_isomorphic(
+            top, isomorphic_top_graphs, forcefields
+        )  # what if only one forcefield?, traversing done in this function
```

Comment:
Can you elaborate on the question posed in this comment?

Reply:
This question was a note to myself to handle the case that only one forcefield is passed, without specifying the molecules it should be attached to. Currently, that will just spit out an error since it expects `forcefields` to be of type dict. So we should either tell users in this case to use identify_connected_components=False, or just automatically use the single forcefield that was passed.